src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/DF_BaseRecalibrator.java

```diff
@@ -76,68 +84,134 @@
-    protected String project;
+    // set up with the user's arguments, kept around to save the textual report.
+    private BaseRecalibratorUprooted baseRecalibratorUprooted;
 
-    // not actually optional if you ask for remote exec
-    @Argument(doc = "Google Cloud Storage path for intermediate files", optional = true)
-    protected String stagingLocation;
+    @Override
+    protected void setupPipeline(Pipeline pipeline) {
+        try {
+            remote = runnerType != PipelineRunnerType.LOCAL;
+            ingestInputs(pipeline);
+            BQSR_Dataflow.ensureReferenceIsReadable(pipeline.getOptions(), referencePath);
+            baseRecalibratorUprooted = BaseRecalibratorUprooted.fromArgs(header, BRAC);
+            baseRecalibratorUprooted.checkClientArguments();
 
-    @Argument(doc = "Number of Dataflow workers", optional = true)
-    protected int numWorkers = 0;
+            // 2. set up computation
+            PCollection<RecalibrationTables> aggregated =
+                    BQSR_Dataflow.GetRecalibrationTables(header, reads, referencePath, BRAC, skipIntervals);
 
-    // ------------------------------------------
+            if (null == outputTablesPath) {
+                // we need those, so let's pick a location for them.
+                if (remote) {
+                    outputTablesPath = GcsPath.fromUri(stagingLocation).resolve("temp-ds-recaltables.sj").toString();
+                } else {
+                    File outputTables = File.createTempFile("temp-ds-recaltables", ".sj");
+                    outputTables.deleteOnExit();
+                    outputTablesPath = outputTables.getPath();
+                }
+            }
+            boolean isGCS = serializeSingleObject(aggregated, outputTablesPath);
+            if (remote && !isGCS) {
+                throw new GATKException("If running on the cloud, either leave outputTablesPath unset or point it to a GCS location.");
+            }
+        } catch (RuntimeException rx) {
+            throw rx;
+        } catch (Exception x) {
+            throw new RuntimeException(x);
+        }
+    }
 
-    // the inputs to BQSR
-    private PCollection<Read> reads;
-    private PCollection<SimpleInterval> skipIntervals;
-    private SAMFileHeader header;
+    protected void afterPipeline(Pipeline p) {
+
+        //  Get the table back and output it in text form to RAC.RECAL_TABLE.
+        // TODO: if running on the cloud and the output destination is on the cloud, then it's faster to have a worker do it directly, without the file roundtrip.
+        final InputStream tblIn;
+        if (BucketUtils.isCloudStorageUrl(outputTablesPath)) {
+            try {
+                tblIn = Channels.newInputStream(new GcsUtil.GcsUtilFactory().create(p.getOptions()).open(GcsPath.fromUri(outputTablesPath)));
+            } catch (IOException e) {
+                throw new GATKException("Unexpected: unable to read results file. (bug?)", e);
+            }
```

Comment:
Sure thing! I called it "openFile", though, since it can open local files as well.